jetbrains-core/src/software/aws/toolkits/jetbrains/ui/clouddebug/ArtifactMappingPopup.kt

```diff
@@ -0,0 +1,158 @@
+import com.intellij.ui.components.JBLabel
+import com.intellij.ui.popup.list.ListPopupImpl
+import com.intellij.util.ui.JBUI
+import net.miginfocom.swing.MigLayout
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.core.utils.trace
+import software.aws.toolkits.jetbrains.services.ecs.execution.ArtifactMapping
+import software.aws.toolkits.resources.message
+import java.awt.Component
+import javax.swing.Icon
+import javax.swing.JList
+import javax.swing.JPanel
+import javax.swing.ListCellRenderer
+
+/**
+ * Popup represents values from Artifact Mapping table
+ */
+class ArtifactMappingPopup {
+
+    companion object {
+        private const val MAX_ROW_COUNT = 5
+
+        fun createPopup(
+            artifactMappingItems: List<ArtifactMapping>,
+            onSelected: (ArtifactMapping?) -> Unit
+        ): ListPopup {
+            val step = ArtifactMappingPopupStep(artifactMappingItems, onSelected)
+            val popup = JBPopupFactory.getInstance().createListPopup(step, MAX_ROW_COUNT)
+
+            val popupToUpdate = popup as? ListPopupImpl
+                ?: throw IllegalStateException("Unable to cast popup in type: '${popup.javaClass}' to ${ListPopupImpl::class.qualifiedName}")
+
+            popupToUpdate.list.setCellRenderer(PathMappingPopupCellRenderer())
+            return popupToUpdate
+        }
+    }
+}
+
+class ArtifactMappingPopupStep(paths: List<ArtifactMapping>, private val onSelected: (ArtifactMapping?) -> Unit) :
+    BaseListPopupStep<ArtifactMapping>(message("cloud_debug.run_configuration.generate_automatically.popup_title"), paths) {
+
+    override fun isSpeedSearchEnabled(): Boolean = true
+
+    override fun hasSubstep(selectedValue: ArtifactMapping?): Boolean = false
+
+    override fun getIconFor(value: ArtifactMapping?): Icon? = null
+
+    override fun onChosen(value: ArtifactMapping?, finalChoice: Boolean): PopupStep<*>? {
+        val stepResult = super.onChosen(value, finalChoice)
+        onSelected(value)
+        return stepResult
+    }
+
+    override fun isAutoSelectionEnabled(): Boolean = false
+}
+
+/**
+ * Renderer for a popup cell.
+ * A popup contains a line presentation for mapping between local path and remote path in container.
+ * A renderer present this information using the following layout: {{local_path} > {remote_path}}, where
+ * {local_path} has min fixed width for every element in a popup to align all values.
+ * Width for {local_path} component is set to a maximum width of existing components, but not more then a defined maximum.
+ */
+class PathMappingPopupCellRenderer : CellRendererPanel(), ListCellRenderer<ArtifactMapping> {
+
+    companion object {
+        private val logger = getLogger<PathMappingPopupCellRenderer>()
+
+        private const val LEFT_COMPONENT_MAX_WIDTH = 300
+        private const val LEFT_COMPONENT_MIN_WIDTH = 50
+    }
+
+    private var leftComponentWidth: Int = -1
+    private val localPathLabel = JBLabel()
```

Comment:
Do we need 3 labels or can we just combine the string into 1 label with string concatenation?

Or am I reading this that we will vertically align everything in the list so that all the separators are vertically aligned?